src/scheduler/gc_works.rs

```diff
@@ -0,0 +1,462 @@
+        <E::VM as VMBinding>::VMScanning::notify_initial_thread_scan_complete(false, worker.tls);
+        mmtk.plan.common().base.set_gc_status(GcStatus::GcProper);
+    }
+}
+
+pub struct ScanStackRoot<Edges: ProcessEdgesWork>(
+    pub &'static mut Mutator<SelectedPlan<Edges::VM>>,
+);
+
+impl<E: ProcessEdgesWork> GCWork<E::VM> for ScanStackRoot<E> {
+    fn do_work(&mut self, worker: &mut GCWorker<E::VM>, mmtk: &'static MMTK<E::VM>) {
+        trace!("ScanStackRoot for mutator {:?}", self.0.get_tls());
+        <E::VM as VMBinding>::VMScanning::scan_thread_root::<E>(
+            unsafe { &mut *(self.0 as *mut _) },
+            worker.tls,
+        );
+        self.0.flush();
+        let old = SCANNED_STACKS.fetch_add(1, Ordering::SeqCst);
+        if old + 1 == <E::VM as VMBinding>::VMActivePlan::number_of_mutators() {
+            SCANNED_STACKS.store(0, Ordering::SeqCst);
+            <E::VM as VMBinding>::VMScanning::notify_initial_thread_scan_complete(
+                false, worker.tls,
+            );
+            mmtk.plan.common().base.set_gc_status(GcStatus::GcProper);
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct ScanVMSpecificRoots<Edges: ProcessEdgesWork>(PhantomData<Edges>);
+
+impl<E: ProcessEdgesWork> ScanVMSpecificRoots<E> {
+    pub fn new() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<E: ProcessEdgesWork> GCWork<E::VM> for ScanVMSpecificRoots<E> {
+    fn do_work(&mut self, _worker: &mut GCWorker<E::VM>, _mmtk: &'static MMTK<E::VM>) {
+        trace!("ScanStaticRoots");
+        <E::VM as VMBinding>::VMScanning::scan_vm_specific_roots::<E>();
+    }
+}
+
+pub struct ProcessEdgesBase<E: ProcessEdgesWork> {
+    pub edges: Vec<Address>,
+    pub nodes: Vec<ObjectReference>,
+    pub mmtk: Option<&'static MMTK<E::VM>>,
+    worker: *mut GCWorker<E::VM>,
+}
+
+unsafe impl<E: ProcessEdgesWork> Sync for ProcessEdgesBase<E> {}
+unsafe impl<E: ProcessEdgesWork> Send for ProcessEdgesBase<E> {}
+
+impl<E: ProcessEdgesWork> Default for ProcessEdgesBase<E> {
+    fn default() -> Self {
+        Self {
+            edges: vec![],
+            nodes: vec![],
+            mmtk: None,
+            worker: 0 as _,
+        }
+    }
+}
+
+impl<E: ProcessEdgesWork> ProcessEdgesBase<E> {
+    pub fn new(edges: Vec<Address>) -> Self {
+        Self {
+            edges,
+            ..Self::default()
+        }
+    }
+    pub fn set_worker(&mut self, worker: &mut GCWorker<E::VM>) {
+        self.worker = worker;
+    }
+    #[inline]
+    pub fn worker(&self) -> &'static mut GCWorker<E::VM> {
+        unsafe { &mut *self.worker }
+    }
+    #[inline]
+    pub fn mmtk(&self) -> &'static MMTK<E::VM> {
+        self.mmtk.unwrap()
+    }
+    #[inline]
+    pub fn plan(&self) -> &'static SelectedPlan<E::VM> {
+        &self.mmtk.unwrap().plan
+    }
+}
+
+/// Scan & update a list of object slots
+pub trait ProcessEdgesWork:
+    Send + Sync + 'static + Sized + DerefMut + Deref<Target = ProcessEdgesBase<Self>>
```

Comment:
Not sure if this is an established pattern (or even a good pattern). But this can reduces the frequent use of `self.base().xxx` and can simplifies the code a lot. Probably I'll change to `self.base()` to keep the style aligned with the `Plan`/`BasePlan` code.

Reply:
I am not feeling negative about it. It just made me wonder why it derefs to a base type. Once I figured out it was used to access fields from the base type, it makes sense.